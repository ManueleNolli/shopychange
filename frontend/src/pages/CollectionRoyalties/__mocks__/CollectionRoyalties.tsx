import { divWithoutChildrenMock } from '../../../../mocksUtils'

const CollectionRoyaltiesMock = () =>
  divWithoutChildrenMock('CollectionRoyalties')
export default CollectionRoyaltiesMock
