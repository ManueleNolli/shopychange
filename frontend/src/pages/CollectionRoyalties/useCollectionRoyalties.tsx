{
  /* eslint-disable */
}
import { useEffect, useState } from 'react'
import { useUserContext } from '../../context/userContext'
import useFetching from '../../hooks/useFetching'
import { RoyaltyCollection } from '../../types/components/RoyaltyCollection'
import { useToast } from '@chakra-ui/react'
import {
  collectionRoyaltyQuery,
  handlerQueryErrorCatch,
} from '../../services/BackendService'
import {
  updateCollectionRoyalties,
  waitTransaction,
  resetDefaultCollectionRoyalties,
} from '../../services/BlockchainService'
import useLoading from '../../hooks/useLoading'
import { FormValues } from './CollectionRoyalties'
import { Contract } from '../../types/components/Contract'

export default function useCollectionRoyalties(address: string) {
  const toast = useToast()
  const { userAddress, blockchainNetworkId } = useUserContext()
  const { isFetching, setIsFetching, isError, setIsError } = useFetching()

  const { isLoading: isUpdateLoading, setIsLoading: setIsUpdateLoading } =
    useLoading()
  const {
    isLoading: isRemoveDefaultLoading,
    setIsLoading: setIsRemoveDefaultLoading,
  } = useLoading()
  const [collection, setCollection] = useState<Contract>()
  const [royalties, setRoyalties] = useState<RoyaltyCollection>()

  const fetchRoyalties = async () => {
    if (!userAddress || !blockchainNetworkId) return

    const response = await handlerQueryErrorCatch(
      collectionRoyaltyQuery(blockchainNetworkId, address)
    )

    if (response.error) {
      setIsError(true)
      setIsFetching(false)
      return
    }

    const result = response.result
    setCollection(result.contract)
    setRoyalties(result.royaltyCollection)
  }

  useEffect(() => {
    fetchRoyalties().then(() => {
      setIsFetching(false)
    })
  }, [address])

  const handleSubmit = async (values: FormValues) => {
    setIsUpdateLoading(true)
    updateCollectionRoyalties(address, values.royalties)
      .then(waitTransaction)
      .then(() => {
        setIsUpdateLoading(false)
        fetchRoyalties().then(() => {
          toast({
            variant: 'solid',
            title: 'Transaction successful',
            description: 'Royalties updated',
            status: 'success',
            duration: 9000,
            isClosable: true,
          })
        })
      })
      .catch((err: any) => {
        // First phrase of the error message is enough
        const errorMessage = err.message.split('.')[0]
        toast({
          variant: 'solid',
          title: 'Transaction failed',
          description: errorMessage,
          status: 'warning',
          duration: 9000,
          isClosable: true,
        })
        setIsUpdateLoading(false)
      })
  }

  const handleRemoveDefault = async () => {
    setIsRemoveDefaultLoading(true)
    resetDefaultCollectionRoyalties(address)
      .then(waitTransaction)
      .then(() => {
        setIsRemoveDefaultLoading(false)
        fetchRoyalties().then(() => {
          toast({
            variant: 'solid',
            title: 'Transaction successful',
            description: 'Default royalties removed',
            status: 'success',
            duration: 9000,
            isClosable: true,
          })
        })
      })
      .catch((err: any) => {
        // First phrase of the error message is enough
        const errorMessage = err.message.split('.')[0]
        toast({
          variant: 'solid',
          title: 'Transaction failed',
          description: errorMessage,
          status: 'warning',
          duration: 9000,
          isClosable: true,
        })
        setIsRemoveDefaultLoading(false)
      })
  }

  return {
    isFetching,
    isError,
    isUpdateLoading,
    isRemoveDefaultLoading,
    collection,
    royalties,
    handleSubmit,
    handleRemoveDefault,
  }
}
