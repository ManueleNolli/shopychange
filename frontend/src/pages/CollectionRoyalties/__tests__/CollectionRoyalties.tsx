import React from 'react'
import { render, act, fireEvent, screen } from '@testing-library/react'
import CollectionRoyalties from '../CollectionRoyalties'
import useCollectionRoyalties from '../useCollectionRoyalties'

jest.mock('../useCollectionRoyalties', () => jest.fn())

jest.mock('../../../components/WithdrawRoyaltyButton/WithdrawRoyaltyButton')
jest.mock('../../../utils/SimplifiedAddress/SimplifiedAddress', () => ({
  renderAddress: jest.fn((address: string) => address),
}))
describe('CollectionRoyalties', () => {
  test('render successfully', () => {
    ;(useCollectionRoyalties as jest.Mock).mockReturnValue({
      isFetching: false,
      isError: false,
      isUpdateLoading: false,
      isRemoveDefaultLoading: false,
      collection: {
        address: '0x00',
        name: 'name',
        symbol: 'symbol',
      },
      royalties: {
        royalties: [
          {
            recipient: '0x00',
            share: 10,
          },
        ],
        royaltySum: 10,
        supportsErc2981: true,
        supportsErc2981MultiReceiver: true,
        hasPaymentSplitter: true,
      },
      handleSubmit: jest.fn(),
      handleRemoveDefault: jest.fn(),
    })
    const { container } = render(<CollectionRoyalties address="0x00" />)
    expect(container).toMatchSnapshot()
  })

  test('validate address', async () => {
    ;(useCollectionRoyalties as jest.Mock).mockReturnValue({
      isFetching: false,
      isError: false,
      isUpdateLoading: false,
      isRemoveDefaultLoading: false,
      collection: {
        address: '0x00',
        name: 'name',
        symbol: 'symbol',
      },
      royalties: {
        royalties: [
          {
            recipient: '0x00',
            share: 10,
          },
        ],
        royaltySum: 10,
        supportsErc2981: true,
        supportsErc2981MultiReceiver: true,
        hasPaymentSplitter: true,
      },
      handleSubmit: jest.fn(),
      handleRemoveDefault: jest.fn(),
    })
    const tree = render(<CollectionRoyalties address="0x00" />)

    const button = screen.getByLabelText('add-royalty')

    // act to add address
    await act(async () => {
      fireEvent.click(button)
    })

    const input = tree.getAllByPlaceholderText('Receiver')[1]

    await act(async () => {
      fireEvent.change(input, { target: { value: '0x00' } })
    })

    expect(input).toHaveValue('0x00')
  })

  test('change value number', async () => {
    ;(useCollectionRoyalties as jest.Mock).mockReturnValue({
      isFetching: false,
      isError: false,
      isUpdateLoading: false,
      isRemoveDefaultLoading: false,
      collection: {
        address: '0x00',
        name: 'name',
        symbol: 'symbol',
      },
      royalties: {
        royalties: [
          {
            recipient: '0x00',
            share: 10,
          },
        ],
        royaltySum: 10,
        supportsErc2981: true,
        supportsErc2981MultiReceiver: true,
        hasPaymentSplitter: true,
      },
      handleSubmit: jest.fn(),
      handleRemoveDefault: jest.fn(),
    })
    const tree = render(<CollectionRoyalties address="0x00" />)

    const input = tree.getByPlaceholderText('Share (%)')
    await act(async () => {
      fireEvent.change(input, { target: { value: '15' } })
    })

    expect(input).toHaveValue('15')
  })

  test('add address', async () => {
    ;(useCollectionRoyalties as jest.Mock).mockReturnValue({
      isFetching: false,
      isError: false,
      isUpdateLoading: false,
      isRemoveDefaultLoading: false,
      collection: {
        address: '0x00',
        name: 'name',
        symbol: 'symbol',
      },
      royalties: {
        royalties: [
          {
            recipient: '0x00',
            share: 10,
          },
        ],
        royaltySum: 10,
        supportsErc2981: true,
        supportsErc2981MultiReceiver: true,
        hasPaymentSplitter: true,
      },
      handleSubmit: jest.fn(),
      handleRemoveDefault: jest.fn(),
    })
    render(<CollectionRoyalties address="0x00" />)

    const addresses = screen.getAllByPlaceholderText('Receiver')

    expect(addresses.length).toBe(1)

    const button = screen.getByLabelText('add-royalty')

    // act to add address
    await act(async () => {
      fireEvent.click(button)
    })

    const addressesAfterAdd = screen.getAllByPlaceholderText('Receiver')

    expect(addressesAfterAdd.length).toBe(2)
  })
  test('remove address', async () => {
    ;(useCollectionRoyalties as jest.Mock).mockReturnValue({
      isFetching: false,
      isError: false,
      isUpdateLoading: false,
      isRemoveDefaultLoading: false,
      collection: {
        address: '0x00',
        name: 'name',
        symbol: 'symbol',
      },
      royalties: {
        royalties: [
          {
            recipient: '0x00',
            share: 10,
          },
        ],
        royaltySum: 10,
        supportsErc2981: true,
        supportsErc2981MultiReceiver: true,
        hasPaymentSplitter: true,
      },
      handleSubmit: jest.fn(),
      handleRemoveDefault: jest.fn(),
    })
    render(<CollectionRoyalties address="0x00" />)

    const addresses = screen.getAllByPlaceholderText('Receiver')

    expect(addresses.length).toBe(1)

    const button = screen.getByLabelText('remove-royalty')

    // act to add address
    await act(async () => {
      fireEvent.click(button)
    })

    const addressesAfterAdd = screen.queryAllByPlaceholderText('Receiver')

    expect(addressesAfterAdd.length).toBe(0)
  })
})
