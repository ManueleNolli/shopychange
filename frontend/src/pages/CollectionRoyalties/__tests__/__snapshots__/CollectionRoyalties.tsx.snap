// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CollectionRoyalties render successfully 1`] = `
<div>
  <div
    class="css-0"
  >
    <div
      class="css-b93e9j"
    >
      <h2
        class="chakra-heading css-0"
      >
        Royalties 
        name
      </h2>
      <h2
        class="chakra-heading css-odz94x"
      />
    </div>
    <div
      class="css-wjed84"
    >
      <div
        class="css-0"
      >
        <form
          action="#"
        >
          <div
            class="css-fe3jho"
          >
            <label
              class="chakra-form__label css-uvho5g"
            >
              Royalties
            </label>
            <div
              class="css-bj7zsy"
            >
              <div
                class="css-1so8qpg"
              >
                <div
                  class="chakra-form-control css-0"
                  role="group"
                >
                  <input
                    class="chakra-input css-0"
                    id="royalties[0].receiver"
                    name="royalties[0].receiver"
                    placeholder="Receiver"
                    value=""
                  />
                </div>
              </div>
              <div
                class="css-17xejub"
              />
              <div
                class="css-18lv6ah"
              >
                <div
                  class="chakra-form-control css-0"
                  role="group"
                >
                  <div
                    class="chakra-numberinput css-e69dqy"
                    value="10"
                  >
                    <input
                      aria-readonly="false"
                      aria-required="false"
                      aria-valuemax="100"
                      aria-valuemin="0"
                      aria-valuenow="10"
                      aria-valuetext="10"
                      autocomplete="off"
                      autocorrect="off"
                      class="chakra-numberinput__field css-8atqhb"
                      id="royalties[0].share"
                      inputmode="decimal"
                      name="royalties[0].share"
                      pattern="[0-9]*(.[0-9]+)?"
                      placeholder="Share (%)"
                      role="spinbutton"
                      type="text"
                      value="10"
                    />
                  </div>
                </div>
              </div>
              <div
                class="css-a6plov"
              >
                <div
                  class="css-cse5we"
                >
                  <svg
                    aria-label="remove-royalty"
                    fill="#FF4500"
                    height="80%"
                    stroke="currentColor"
                    stroke-width="0"
                    viewBox="0 0 24 24"
                    width="80%"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M0 0h24v24H0z"
                      fill="none"
                    />
                    <path
                      d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
                    />
                  </svg>
                </div>
              </div>
            </div>
            <div
              class="css-98x68k"
            >
              <div
                class="css-913hcs"
              >
                <svg
                  aria-label="add-royalty"
                  baseProfile="tiny"
                  fill="#228B22"
                  height="100%"
                  stroke="currentColor"
                  stroke-width="0"
                  version="1.2"
                  viewBox="0 0 24 24"
                  width="100%"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M18 10h-4v-4c0-1.104-.896-2-2-2s-2 .896-2 2l.071 4h-4.071c-1.104 0-2 .896-2 2s.896 2 2 2l4.071-.071-.071 4.071c0 1.104.896 2 2 2s2-.896 2-2v-4.071l4 .071c1.104 0 2-.896 2-2s-.896-2-2-2z"
                  />
                </svg>
              </div>
            </div>
          </div>
          <div
            class="css-b93e9j"
          >
            <i
              class="chakra-text css-mp67vj"
            >
              Make sure to withdraw all the funds before updating. If you don't, the funds may be lost
            </i>
            <button
              class="chakra-button css-adqxy3"
              data-testid="onClick-button"
              style="white-space: initial;"
              type="submit"
            >
              Update Royalties
            </button>
          </div>
        </form>
        <hr
          aria-orientation="horizontal"
          class="chakra-divider css-1lv7lm1"
        />
        <div
          class="css-b93e9j"
        >
          <h2
            class="chakra-heading css-1jx9xo4"
          >
            Other actions
          </h2>
          <button
            class="chakra-button css-1vu7gvf"
            data-testid="onClick-button"
            style="white-space: initial;"
            type="button"
          >
            Remove default royalty
          </button>
          <div
            data-testid="WithdrawRoyaltyButton"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
